django_pg_simple_hll/migrations/0002_custom_hashing.sql

```diff
@@ -0,0 +1,186 @@
+--  it use used to calculate the number of buckets
+CREATE OR REPLACE FUNCTION hll_bucket(
+    hll_agg_state int [],
+    hashed_input int,
+    hll_precision int
+) RETURNS int []
+LANGUAGE plpgsql IMMUTABLE STRICT PARALLEL SAFE AS $$
+DECLARE
+    -- eg. a hll_precision 9 sets the number of buckets to 512, because 2**9 -> 512
+    n_buckets int := POW(2, hll_precision);
+    -- bucket the hash into one of n buckets
+    -- we add 1 because postgres arrays are 1-indexed
+    bucket_key int := (hashed_input & (n_buckets - 1)) + 1;
+    -- length of current state array
+    hll_agg_state_length int := ARRAY_LENGTH(hll_agg_state, 1);
+    -- pre fetch the hash for occupying the corresponding bucket for the input
+    current_hash int := hll_agg_state[bucket_key];
+BEGIN
+    -- we can only handle precision up to 26 or 67,108,864 buckets
+    -- because array size is limited to 134,217,727 (or 2^27 - 1) in postgres
+    IF hll_precision < 4 OR hll_precision > 26 THEN
+        RAISE EXCEPTION 'invalid hll_precision: % - must be between 4 (16 buckets) and 26 (67,108,864 buckets) inclusive',
+            hll_precision;
+    END IF;
+    -- postgres squeezes uninitialised elements in an array,
+    -- so we add a first and last NULL elements to keep a fixed size
+    IF hll_agg_state_length IS NULL OR hll_agg_state_length < n_buckets THEN
+        hll_agg_state[1] := COALESCE(hll_agg_state[1], NULL);
+        hll_agg_state[n_buckets] := COALESCE(hll_agg_state[n_buckets], NULL);
+    END IF;
+
+    IF current_hash IS NULL OR current_hash > hashed_input THEN
+        hll_agg_state[bucket_key] := hashed_input;
+    END IF;
+    RETURN hll_agg_state;
+END $$;
+
+-- hash and bucket in one function
+CREATE OR REPLACE FUNCTION hll_hash_and_bucket(
+    hll_agg_state int [],
+    input anyelement,
+    hll_precision int
+) RETURNS int []
+LANGUAGE sql IMMUTABLE STRICT PARALLEL SAFE AS $$
+SELECT hll_bucket(hll_agg_state, hll_hash(input), hll_precision);
+$$;
+
+-- hash and bucket in one function with default precision of 9
+CREATE OR REPLACE FUNCTION hll_hash_and_bucket(
+    hll_agg_state int [],
+    input anyelement
+) RETURNS int []
+LANGUAGE plpgsql IMMUTABLE STRICT PARALLEL SAFE AS $$
+BEGIN
+    RETURN hll_hash_and_bucket(hll_agg_state, input, 9);
+END $$;
+
+-- The combinefunc
+-- combines two states, taking the smaller hash from each corresponding index
+-- we don't have any extra logic here for the two flanking extra elements at
+-- the beginning and the end of the array because they should just be
+-- copied as they are the same
+CREATE OR REPLACE FUNCTION hll_bucket_combine(
+    hll_left_agg_state int [],
+    hll_right_agg_state int []
+) RETURNS int []
+LANGUAGE sql IMMUTABLE STRICT PARALLEL SAFE AS $$
+SELECT ARRAY(
+    SELECT
+        LEAST(left_bucket_hash, right_bucket_hash)
+    FROM
+        UNNEST(hll_left_agg_state, hll_right_agg_state) AS AGG_STATE(left_bucket_hash, right_bucket_hash)
+) $$;
+
+-- The finalfunc
```

Comment:
This function is copied unmodified over from `0001_initial.sql` so that we have the complete implementation in one file